src/main/java/org/folio/auth/authtokenmodule/tokens/Token.java

```diff
@@ -0,0 +1,248 @@
+  }
+
+  /**
+   * Validate all the things that tokens have in common.
+   * @param request The http request context where the token is being provided.
+   * @throws TokenValidationException Throws this when any validation step fails, but may throw
+   * other exceptions as well.
+   */
+  protected void validateCommon(HttpServerRequest request) throws TokenValidationException {
+    try {
+      // Check that the token has a source.
+      if (source == null)
+        throw new TokenValidationException("Token has no source defined", 500);
+
+      // Check that the claims have been created.
+      if (claims == null)
+        throw new TokenValidationException("Token has no claims", 500);
+
+      // Check some claims that all tokens must have.
+      String[] requiredClaims = new String[] { "sub", "tenant", "type" };
+      for (String c : requiredClaims) {
+        if (!claims.containsKey(c)) {
+          throw new TokenValidationException(String.format("Token is missing %s claim", c), 500);
+        }
+      }
+
+      if (request == null)
+        return;
+
+      // Check that some items in the headers match what are in the token.
+      String headerTenant = request.headers().get(XOkapiHeaders.TENANT);
+      if (!claims.getString("tenant").equals(headerTenant))
+        throw new TokenValidationException("Tenant in header does not equal tenant in token", 403);
+
+      String headerUserId = request.headers().get(XOkapiHeaders.USER_ID);
+      if (headerUserId != null && !claims.getString("user_id").equals(headerUserId))
+        throw new TokenValidationException("User id in header does not equal user id in token", 403);
+
+    } catch (TokenValidationException e) {
+      throw e;
+    } catch (Exception e) {
+      throw new TokenValidationException("Unexpected token validation exception", e, 500);
+    }
+  }
+
+  private static Token parse(String sourceToken, TokenCreator tokenCreator) throws TokenValidationException {
+    Token token = null;
+    JsonObject claims = null;
+    final String invalidTokenMsg = "Invalid token";
+
+    try {
+      if (isEncrypted(sourceToken)) {
+        String tokenContent = tokenCreator.decodeJWEToken(sourceToken);
+        claims = new JsonObject(tokenContent);
+      } else {
+        tokenCreator.checkJWTToken(sourceToken);
+        claims = getClaims(sourceToken);
+      }
+    } catch (ParseException p) {
+      throw new TokenValidationException(invalidTokenMsg, p, 401);
+    } catch (JOSEException j) {
+      throw new TokenValidationException(invalidTokenMsg, j, 401);
+    } catch (BadSignatureException b) {
+      throw new TokenValidationException(invalidTokenMsg, b, 401);
+    }
+
+    String tokenType = claims.getString("type");
+    if (tokenType == null)
+      throw new TokenValidationException("Token has no type", 400);
+
+    switch (tokenType) {
+      case AccessToken.type:
+        token = new AccessToken(sourceToken, claims);
```

Comment:
Agree it would be less code but then callers have to deal with the potential of a null when a token has no type. Felt it was better to have this method handle that. I was imagining a situation when we would have some tokens "out there" that didn't have types and have a clear message when that occurs.

Reply:
A constructor never returns null. "The value of a class instance creation expression is a reference to the newly created object of the specified class." https://docs.oracle.com/javase/specs/jls/se7/html/jls-15.html#jls-15.9.4

    switch (tokenType) {
      case AccessToken.type:
        return new AccessToken(sourceToken, claims);
      case RefreshToken.type:
        return new RefreshToken(sourceToken, claims);
      case ApiToken.type:
        return new ApiToken(sourceToken, claims);
      case DummyToken.type:
        return new DummyToken(sourceToken, claims);
      case ModuleToken.type:
        return new ModuleToken(sourceToken, claims);
      default:
        throw new TokenValidationException("Unable to parse token", 400);
    }

Reply:
You are right. For some reason I was focused on the type being null, but I check for that above.